File: C++/stacks_and_queues/queue_via_stacks.cpp
#include <iostream>
#include <stack>

/**
 * @brief A queue implemented using two stacks
 * @tparam T the type of the elements in the queue
 */
template <typename T>
class MyQueue {
public:
    /**
     * @brief Adds an element to the back of the queue
     * @tparam U the type of the value being added
     * @param value the value to be added
     */
    template <typename U>
    void add(U &&value) {
        // Push the value onto the newValues stack
        newValues.push(std::forward<U>(value));
    }

    /**
     * @brief Returns the front element of the queue without removing it
     * 
     * If the reversed stack is empty, all the elements in the newValues stack
     * are moved to the reversed stack before the front element is returned.
     * 
     * @return the front element of the queue
     */
    T peek() {
        if (reversed.empty())
            // Move all the elements from the newValues stack to the reversed stack
            move(newValues, reversed);
        // Return the top element of the reversed stack
        return reversed.top();
    }

    /**
     * @brief Removes and returns the front element of the queue
     * 
     * If the reversed stack is empty, all the elements in the newValues stack
     * are moved to the reversed stack before the front element is removed and returned.
     * 
     * @return the front element of the queue
     */
    T remove() {
        if (reversed.empty())
            // Move all the elements from the newValues stack to the reversed stack
            move(newValues, reversed);
        // Pop and return the top element of the reversed stack
        T ret = reversed.top();
        reversed.pop();
        return ret;
    }

    /**
     * @brief Returns whether the queue is empty
     * @return true if the queue is empty, false otherwise
     */
    bool isEmpty() {
        // The queue is empty if both stacks are empty
        return newValues.empty() && reversed.empty();
    }

private:
    /**
     * @brief Moves all the elements from one stack to another
     * 
     * This method moves all the elements from the "from" stack to the "to" stack.
     * It does so by popping elements from the "from" stack and pushing them onto
     * the "to" stack until the "from" stack is empty.
     * 
     * @param from the stack to move elements from
     * @param to the stack to move elements to
     */
    static void move(std::stack<T> &from, std::stack<T> &to) {
        while (!from.empty()) {
            // Pop the top element from the "from" stack
            T elem = from.top();
            from.pop();
            // Push the popped element onto the "to" stack
            to.push(elem);
        }
    }

    std::stack<T> newValues; /**< The stack for new values */
    std::stack<T> reversed; /**< The stack for reversed values */
};


int main()
{

int main() {
    MyQueue<int> queue; // create a new queue object that holds integers

    for (int i = 0; i < 10; ++i) {
        queue.add(i); // add the integer i to the queue
        std::cout << "Queued value " << i << std::endl; // print a message to the console for each integer that is added
    }

    for (int i = 0; i < 5; ++i)
        std::cout << "Removed value " << queue.remove() << std::endl; // remove the first five integers from the queue and print a message to the console for each integer that is removed

    for (int i = 10; i < 15; ++i) {
        queue.add(i); // add the integer i to the queue
        std::cout << "Queued value " << i << std::endl; // print a message to the console for each integer that is added
    }

    while (!queue.isEmpty())
        std::cout << "Removed value " << queue.remove() << std::endl; // remove all the remaining integers from the queue and print a message to the console for each integer that is removed

    return 1; // return the value 1 to the calling process
}
}

File: Python/stacks_and_queues/queue_via_stacks.py
# 3.4 Queue Via Stacks
import unittest
from stack import Stack

class MyQueue:
    """
    A queue data structure implemented using two stacks.
    """

    def __init__(self):
        """
        Initializes the queue with two empty stacks: new_stack and old_stack.
        """
        self.new_stack = Stack()    # create an empty stack for incoming elements
        self.old_stack = Stack()    # create an empty stack for outgoing elements

    def _shift_stacks(self):
        """
        Internal helper method that shifts elements from new_stack to old_stack,
        so that the oldest elements are always at the top of old_stack.
        """
        if self.old_stack.is_empty():   # if old_stack is empty
            while not self.new_stack.is_empty():   # while new_stack is not empty
                self.old_stack.push(self.new_stack.pop())    # pop an element from new_stack and push it to old_stack

    def add(self, value):
        """
        Adds an element to the end of the queue.

        @param value: The value to be added.
        """
        return self.new_stack.push(value)    # push the value onto new_stack

    def peek(self):
        """
        Returns the oldest element in the queue without removing it.

        @return: The oldest element in the queue, or False if the queue is empty.
        """
        if self.is_empty():    # if the queue is empty
            return False
        self._shift_stacks()    # shift elements from new_stack to old_stack
        return self.old_stack.peek()    # return the top element of old_stack

    def remove(self):
        """
        Removes and returns the oldest element in the queue.

        @return: The oldest element in the queue, or False if the queue is empty.
        """
        if self.is_empty():    # if the queue is empty
            return False
        self._shift_stacks()    # shift elements from new_stack to old_stack
        return self.old_stack.pop()    # pop and return the top element of old_stack

    def is_empty(self):
        """
        Checks if the queue is empty.

        @return: True if the queue is empty, False otherwise.
        """
        return len(self) == 0    # return True if the total length of new_stack and old_stack is 0

    def __len__(self):
        """
        Returns the number of elements in the queue.

        @return: The number of elements in the queue.
        """
        return len(self.new_stack) + len(self.old_stack)    # return the sum of the lengths of new_stack and old_stack


import unittest

class Tests(unittest.TestCase):
    # Define test cases to be used in the tests
    test_cases = [([1, 2, 3]), ([-1, 0, 1]), (["a", "b", "c", "d", "e", "f"])]

    def test_size(self):
        # Test whether add and remove methods correctly adjust the length of the queue
        for sequence in self.test_cases:
            q = MyQueue()
            for index, val in enumerate(sequence, start=1):
                q.add(val)
                # Assert that length of queue is the same as the number of elements added
                assert len(q) == index
            for index, val in enumerate(sequence, start=1):
                q.remove()
                # Assert that length of queue is the original length minus the number of elements removed
                assert len(q) == len(sequence) - index

    def test_add(self):
        # Test whether add method adds elements to the queue and that peek method returns first element
        for sequence in self.test_cases:
            q = MyQueue()
            for val in sequence:
                q.add(val)
            # Assert that the first element in the queue is the same as the first element in the sequence
            assert q.peek() == sequence[0]
            # Assert that the length of the queue is the same as the length of the sequence
            assert len(q) == len(sequence)

    def test_shift_stacks(self):
        # Test whether the MyQueue class correctly shifts elements from the new_stack to the old_stack when necessary
        for sequence in self.test_cases:
            q = MyQueue()
            for val in sequence:
                q.add(val)
            # Assert that the new_stack is empty
            assert len(q.old_stack) == 0
            # Assert that the new_stack is the same length as the sequence
            assert len(q.new_stack) == len(sequence)
            # Assert that the top element in the new_stack is the last element in the sequence
            assert q.new_stack.peek() == sequence[-1]
            q._shift_stacks()
            # Assert that the old_stack is the same length as the sequence
            assert len(q.old_stack) == len(sequence)
            # Assert that the new_stack is now empty
            assert len(q.new_stack) == 0
            # Assert that the top element in the old_stack is the first element in the sequence
            assert q.old_stack.peek() == sequence[0]

    def test_peek(self):
        # Test whether peek method returns the first element in the queue and updates its value correctly
        for sequence in self.test_cases:
            q = MyQueue()
            for val in sequence:
                q.add(val)
                # Assert that the first element in the queue is the same as the first element in the sequence
                assert q.peek() == sequence[0]
            q.remove()
            # Assert that the first element in the queue is now the second element in the sequence
            assert q.peek() == sequence[1]

    def test_remove(self):
        # Test whether remove method correctly removes and returns elements from the queue in the correct order
        for sequence in self.test_cases:
            q = MyQueue()
            for val in sequence:
                q.add(val)
            for i in range(len(sequence)):
                # Assert that the removed element is the same as the current element in the sequence
                assert q.remove() == sequence[i]

    def test_peek_simple(self):
        # Test the functionality of the peek method with a simple queue
        q = MyQueue()
        q.add(4)
        q.add(6)
        # Assert that the first element in the queue is the first element that was added
        assert q.peek() == 4

    def test_add_simple(self):
        # Test the functionality of the add method with a simple queue
        q = MyQueue()
        q.add(4)
        q.add(6)
        # Assert that the first element in the queue is the first element that was added
        assert q.peek() == 4
        q.add(101)
        # Assert that the first element in the queue is still the first element that was added
        assert q.peek() != 101

    def test_remove_simple(self):
        # Test the functionality of the remove method with a simple queue
        q = MyQueue()
        q.add(4)
        q.add(6)
        # Assert that the length of the queue is 2
        assert len(q) == 2
        # Assert that the removed element is the first element that was added
        assert q.remove() == 4
        # Assert that the removed element is the second element that was added
        assert q.remove() == 6
        # Assert that the length of the queue is now 0
        assert len(q) == 0
        # Assert that trying to remove an element from the queue returns None
        assert not q.remove()

if __name__ == '__main__':
    unittest.main()
